enhancement: introduce theme processing (#9133)
owncloud/ocis: services/web/README.md

It's important to note that the feature at the moment is only capable of providing static (js, mjs, e.g.) web applications
and does not support injection of dynamic web applications (custom dynamic backends).

### Loading Themes

Web themes are loaded, if added in the Infinite Scale source code, at build-time from
`<ocis_repo>/services/web/assets/themes`.
This cannot be manipulated at runtime.

Additionally, the administrator can provide custom themes by storing it in the path defined by the environment
variable `WEB_ASSET_THEMES_PATH`.

With the theme root directory defined, the system needs to know which theme to use.
This can be done by setting the `WEB_UI_THEME_PATH` environment variable.

The final theme is composed of the built-in and the custom theme provided by the

> administrator via `WEB_ASSET_THEMES_PATH` and `WEB_UI_THEME_PATH`.

Reviewer: Can we please (!!!) also clean up this env var mess? Nobody understood it, ever, why there are two variables. there has been so much confusion in the community and internally! And the env vars don't even have the same naming scheme (`WEB_ASSET_THEMES_` and `WEB_UI_THEME_`), meh. Let's make it better.

Reviewer: E.g. What about one single env var `OCIS_THEME_ASSETS_PATH` where the relative path and name of the theme.json file needs to follow a hardcoded value and is ignored otherwise?
I'd also like to avoid the `WEB_xyz` prefix in the env var, because we want to make theming a first class citizen of ocis, for all clients. Not only web. It doesn't matter so much that the defaults still live in web. But the env var name shouldn't hint at that in my opinion.

Author: i agree, its not perfect, please talk to @wkloucek if its needed... if not, we can remove it.

It's important to note that the feature at the moment is only capable of providing static (js, mjs, e.g.) web applications
and does not support injection of dynamic web applications (custom dynamic backends).

### Loading Themes

Web themes are loaded, if added in the Infinite Scale source code, at build-time from
`<ocis_repo>/services/web/assets/themes`.
This cannot be manipulated at runtime.

Additionally, the administrator can provide custom themes by storing it in the path defined by the environment
variable `WEB_ASSET_THEMES_PATH`.

> This environment variable defaults to the Infinite Scale base data directory `$OCIS_BASE_DATA_PATH/web/assets/themes`,
> but can be redefined with any path set manually.
> 

Reviewer 2: ```suggestion
```
We can omit this, as it will be part of the envvar description as proposed. The envvar description change applies to both `WEB_ASSET_THEMES_PATH` and `WEB_UI_THEME_PATH`.

> Once the theme root directory is defined, the system needs to know which theme to use.

Reviewer 2: ```suggestion
With the theme root directory defined, the system needs to know which theme to use.
```

This can be done by setting the `WEB_UI_THEME_PATH` environment variable.

The final theme is composed of the built-in and the custom theme provided by the
administrator via `WEB_ASSET_THEMES_PATH` and `WEB_UI_THEME_PATH`.

> For example, Infinite Scale by default contains a ownCloud theme,
> if the administrator then provides a custom theme via the `WEB_ASSET_THEMES_PATH` directory,
> the theme located at `WEB_ASSET_THEMES_PATH/owncloud/themes.json` will be used instead of the built-in one.

Reviewer 2: ```suggestion
For example, Infinite Scale by default contains a built in ownCloud theme.
If the administrator provides a custom theme via the `WEB_ASSET_THEMES_PATH` directory like,
`WEB_ASSET_THEMES_PATH/owncloud/themes.json`, this one will be used instead of the built-in one.
```

> Some theme keys are mandatory,
> for example, the `common.shareRoles` settings, such mandatory keys are injected automatically at runtime if not provided.

Reviewer 2: ```suggestion
Some theme keys are mandatory, like the `common.shareRoles` settings.
Such mandatory keys are injected automatically at runtime if not provided.
```
